### PR TITLE
[Writer] Allow ndrange_2D and ndrange_3D to coexist in the same input module

### PR DESCRIPTION
**Pull request by AndreyE**
The commit initially made by Scot. It doesn't require fixes in the reader.

### CODE REVIEW

File: test/SPIRV/transcoding/BuildNDRange_2.ll [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+%opencl.block = type opaque
+%struct.ndrange_t = type <{ i32, [3 x i64], [3 x i64], [3 x i64] }>
+
+@block_fn.gs = internal constant [2 x i32] [i32 64, i32 4096], align 4
+@block_fn.gs1 = internal constant [3 x i32] [i32 64, i32 64, i32 64], align 4
+
+; Function Attrs: nounwind
+define spir_func void @block_fn(i32 %level, i32 %maxGlobalWorkSize, i32 addrspace(1)* %rnd, i32 addrspace(1)* %res) #0 {
+entry:
+  %level.addr = alloca i32, align 4
```
Comment (bader) [inline test/SPIRV/transcoding/BuildNDRange_2.ll]: Please, reduce the test case to have minimal number of instructions. Preferably 3 calls of built-in function with constant arguments.

File: lib/SPIRV/SPIRVUtil.cpp [hunk 2]
```diff
@@ -1071,11 +1071,11 @@ Value *
 getScalarOrArray(Value *V, unsigned Size, Instruction *Pos) {
   if (!V->getType()->isPointerTy())
     return V;
-  auto GEP = dyn_cast<GetElementPtrInst>(V);
+  auto GEP = isa<ConstantExpr>(V) || isa<GetElementPtrInst>(V) ? cast<User>(V) : nullptr;
```
Comment (bader) [inline lib/SPIRV/SPIRVUtil.cpp]: Is that relevant to ndrange fix?

File: test/SPIRV/transcoding/BuildNDRange_2.ll [hunk 3]
```diff
@@ -0,0 +1,363 @@
+; RUN: llvm-as %s -o %t.bc
+; RUN: llvm-spirv %t.bc -spirv-text -o %t.spv.txt
+; RUN: FileCheck < %t.spv.txt %s --check-prefix=CHECK-SPIRV
+; RUN: llvm-spirv %t.bc -o %t.spv
+; RUN: llvm-spirv -r %t.spv -o %t.rev.bc
+; RUN: llvm-dis %t.rev.bc
+; RUN: FileCheck < %t.rev.ll %s --check-prefix=CHECK-LLVM
+
+; CHECK-SPIRV-DAG: BuildNDRange {{[0-9]+}} {{[0-9]+}} [[GWS:[0-9]+]] [[LWS:[0-9]+]] [[GWO:[0-9]+]]
```
Comment (bader) [inline test/SPIRV/transcoding/BuildNDRange_2.ll]: What's the point of defining patterns w/o using them?

File: lib/SPIRV/OCL20ToSPIRV.cpp [hunk 4]
```diff
@@ -509,7 +509,7 @@ OCL20ToSPIRV::visitCallNDRange(CallInst *CI,
     default:
       assert(0 && "Invalid number of arguments");
     }
-    return getSPIRVFuncName(OpBuildNDRange);
+    return getSPIRVFuncName(OpBuildNDRange, "_" + std::to_string(Len) + "D");
```
Comment (bader) [inline lib/SPIRV/OCL20ToSPIRV.cpp:514]: I don't understand this change. According to description it's done to avoid some collision. There should be no collisions - all the OpenCL built-ins must map to the same "LLVM-SPIR-V" built-in __spirv_BuildNDRange_*.

Comment (AndreyE) [inline lib/SPIRV/OCL20ToSPIRV.cpp:514]: The problem is in ndrange_2D and ndrange_3D. They are mangled the same way but have different argument types (2D and 3D arrays).

File: test/SPIRV/transcoding/BuildNDRange_2.ll (hunk 3 above)
Comment (AndreyE) [inline test/SPIRV/transcoding/BuildNDRange_2.ll]: --check-prefix=CHECK-SPIRV is defined, to my understanding -DAG postfix allows the test do not relay on the order of the patterns

File: lib/SPIRV/SPIRVUtil.cpp (hunk 2 above)
Comment (AndreyE) [inline lib/SPIRV/SPIRVUtil.cpp]: Yes, it is. I'm updated the test there this should be clear.

File: test/SPIRV/transcoding/BuildNDRange_2.ll (hunk 3 above)
Comment (bader) [inline test/SPIRV/transcoding/BuildNDRange_2.ll]: I meant FileCheck variables GWS, LWS and GWO - they are defined, but not used.
Please, add checks for all ndrange instructions - there should be four of them and validate types of arguments.

Comment (AndreyE) [inline test/SPIRV/transcoding/BuildNDRange_2.ll]: Done
